aten/src/ATen/native/mps/OperationUtils.h

```diff
@@ -0,0 +1,196 @@
+  ~MPSGeneratorImpl() = default;
+
+  void set_current_seed(uint64_t seed) override;
+  uint64_t current_seed() const override;
+  uint64_t seed() override;
+  void set_state(const c10::TensorImpl& new_state) override;
+  c10::intrusive_ptr<c10::TensorImpl> get_state() const override;
+  static DeviceType device_type();
+
+private:
+  MPSGeneratorImpl* clone_impl() const override;
+  uint64_t seed_ = default_rng_seed_val;
+};
+
+const Generator& getDefaultMPSGenerator();
+
+void runMPSGraph(
+    MPSStream* mpsStream,
+    MPSGraph* mpsGraph,
+    NSDictionary* feeds,
+    NSDictionary* results);
+
+MPSDataType getMPSDataType(ScalarType scalar_type);
+MPSDataType getMPSScalarType(ScalarType scalar_type);
+std::string getMPSTypeString(ScalarType scalar_type);
+std::string getMPSShapeString(MPSShape* shape);
+std::string getTensorsStringKey(const TensorList tensors);
+double getMPSScalarValue(const Tensor& t);
+std::string getArrayRefString(const IntArrayRef s);
+
+MPSShape* getMPSShape(const Tensor& t);
+MPSShape* getMPSShape(IntArrayRef sizes);
+MPSShape* getMPSShape(c10::MaybeOwned<Tensor> t);
+
+class Placeholder {
+ public:
+  Placeholder() : _placeholder(nullptr), _value(nullptr) {}
+  Placeholder(MPSGraphTensor* mpsGraphTensor) : _placeholder(mpsGraphTensor), _value(nullptr) {}
+  Placeholder(MPSGraphTensor* mpsGraphTensor, const Tensor& self, MPSShape *mpsShape = nullptr);
+  MPSGraphTensor* getMPSGraphTensor() {
+    return _placeholder;
+  }
+  MPSGraphTensorData* getMPSGraphTensorData() {
+    return _value;
+  }
+  bool isIntermediate() {
+    return _value == nullptr;
+  }
+
+ private:
+  MPSGraphTensor* _placeholder;
+  MPSGraphTensorData* _value;
+};
+
+void resize_tensor(Tensor* output);
+MPSGraphTensor* trunc_tensor(MPSGraph* mpsGraph, MPSGraphTensor* inputTensor);
+MPSGraphTensorData *getMPSGraphTensorData(MPSGraph* mpsGraph,
+                                          MPSStream* mpsStream,
+                                          const Tensor& tensor);
+
+MPSGraph* make_mps_graph();
+void printTensorNDArray(const Tensor& t);
+
+MPSGraphTensor* mpsGraphUnrankedPlaceHolder(MPSGraph *mpsGraph, MPSDataType dataType);
+MPSGraphTensor* mpsGraphRankedPlaceHolder(MPSGraph *mpsGraph, MPSDataType dataType, MPSShape* mpsShape);
+MPSGraphTensor* mpsGraphRankedPlaceHolder(MPSGraph *mpsGraph, const Tensor& tensor);
+MPSGraphTensor* mpsGraphConstantFloatPlaceHolder(MPSGraph *mpsGraph, const double value, MPSShape* mpsShape);
+
+string get_mem_format_string(c10::MemoryFormat memory_format);
+
+using MPSCacheKey = int64_t;
+
+// derive this class to cache a graph and its inputs/ouputs
+// can be used to store any NSObject
+struct MPSCachedGraph
+{
+  MPSCachedGraph(NSObject *object) : _object([object retain]) {}
+  virtual ~MPSCachedGraph() {
+   [_object release];
+   _object = nullptr;
+  }
+  MPSGraph *graph() const { return (MPSGraph *)_object; }
+  NSObject *object() const { return _object; }
+private:
+  NSObject *_object = nullptr;
+};
```

Comment:
A graph cache mmmm

Reply:
The Graph cache is needed to avoid the re-compilation overhead which is significant.